certora/specs/UniversalRewardsDistributor.spec

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+using MerkleTree as MerkleTree;
+using Util as Util;
+
+methods {
+    function root() external returns bytes32 envfree;
+    function claimed(address, address) external returns(uint256) envfree;
+    function claim(address, address, uint256, bytes32[]) external returns(uint256) envfree;
+
+    function MerkleTree.getValue(address, address) external returns(uint256) envfree;
+    function MerkleTree.getHash(bytes32) external returns(bytes32) envfree;
+    function MerkleTree.wellFormedPath(bytes32, bytes32[]) external envfree;
+
+    function _.balanceOf(address) external => DISPATCHER(true);
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+
+    function Util.balanceOf(address, address) external returns(uint256) envfree;
+}
+
+// Check an account claimed amount is correctly updated.
+rule updatedClaimedAmount(address account, address reward, uint256 claimable, bytes32[] proof) {
+    claim(account, reward, claimable, proof);
+
+    assert claimable == claimed(account, reward);
+}
+
+// Check an account can only claim greater amounts each time.
+rule increasingClaimedAmounts(address account, address reward, uint256 claimable, bytes32[] proof) {
+    uint256 claimed = claimed(account, reward);
+
+    claim(account, reward, claimable, proof);
+
+    assert claimable > claimed;
+}
+
+// Check that claiming twice is equivalent to claiming once with the last amount.
+rule claimTwice(address account, address reward, uint256 claim1, uint256 claim2) {
+    storage initStorage = lastStorage;
+
+    bytes32[] proof1; bytes32[] proof2;
+    claim(account, reward, claim1, proof1);
+    claim(account, reward, claim2, proof2);
```

Comment:
It's not possible, but the solver does not see this because it does not know that the Merkle tree is well formed. We can remove it, it's not that interesting now that there is only one root

Reply:
ok, then maybe yes let's remove it